File: android-reference/app/src/main/java/com/echo/twin/tools/RouterAgent.kt
package com.echo.twin.tools

/**
 * Direct port of src/lib/agents/routerAgent.ts. Deliberately unchanged from Round 1
 * to Round 2 — see that file's comment for why intent classification stays
 * deterministic rather than becoming a model call: every additional inference pass
 * on a phone costs real, user-visible seconds, and picking one of five fixed
 * intents doesn't need a language model to do reliably.
 */
enum class Intent {
    FIRST_CHANGE,
    SIMILAR_PERIOD,
    FORENSICS,
    EVIDENCE,
    WHAT_CHANGED,
}

object RouterAgent {
    fun classifyIntent(question: String): Intent {
        val q = question.lowercase()
        return when {
            q.contains("changed first") || q.contains("first change") || q.contains("earliest") -> Intent.FIRST_CHANGE
            q.contains("happened before") || q.contains("similar") || q.contains("before") -> Intent.SIMILAR_PERIOD
            q.contains("why") || q.contains("flagged") || q.contains("forensic") -> Intent.FORENSICS
            q.contains("evidence") -> Intent.EVIDENCE
            else -> Intent.WHAT_CHANGED
        }
    }
}

File: src/lib/agents/routerAgent.ts
// ROUTER AGENT
// Classifies an incoming question and decides which other agent(s) it needs.
// Deterministic by design: on a phone, every round of "let the model decide what to
// do next" costs a full inference pass. Routing is fast, free, and — for a fixed set
// of investigation intents — more reliable than asking a 2-4B model to pick a tool
// name correctly every time. The model's job starts after routing, at Synthesis.

export type Intent = "first_change" | "similar_period" | "forensics" | "evidence" | "what_changed";

export function classifyIntent(question: string): Intent {
  const q = question.toLowerCase();
  if (q.includes("changed first") || q.includes("first change") || q.includes("earliest")) return "first_change";
  if (q.includes("happened before") || q.includes("similar") || q.includes("before")) return "similar_period";
  if (q.includes("why") || q.includes("flagged") || q.includes("forensic")) return "forensics";
  if (q.includes("evidence")) return "evidence";
  return "what_changed";
}
